packages/dds/tree/src/simple-tree/discrepancies.ts

```diff
@@ -321,87 +493,74 @@ function* trackObjectNodeDiscrepancies(
+		} else {
+			yield* getFieldDiscrepancies(
+				fieldSchema,
+				storedSchema,
+				identifier,
+				fieldKey,
+				viewKindIsSequence,
+			);
 		}
 	}
-}
 
-function* compareMaps<K, V1, V2>(
-	a: ReadonlyMap<K, V1>,
-	b: ReadonlyMap<K, V2>,
-): Iterable<
-	| { type: "aExtra"; key: K; value: V1 }
-	| { type: "bExtra"; key: K; value: V2 }
-	| { type: "both"; key: K; valueA: V1; valueB: V2 }
-> {
-	for (const [key, valueA] of a) {
-		const valueB = b.get(key);
-		yield valueB === undefined
-			? { type: "aExtra", key, value: valueA }
-			: { type: "both", key, valueA, valueB };
-	}
-	for (const [key, valueB] of b) {
-		if (!a.has(key)) {
-			yield { type: "bExtra", key, value: valueB };
+	for (const [fieldKey, schema] of stored.objectNodeFields) {
+		// If the stored schema has a field that's not in the view schema
+		if (!viewKeys.has(fieldKey)) {
+			if (schema.kind === forbiddenFieldKindIdentifier) {
+				// In the stored schema the field is explicitly forbidden.
+				// This has the same semantics of the field not being mentioned in the stored schema,
+				// and thus is compatible with the view schema which does not mention this field.
+				continue;
+			}
+			yield {
+				identifier,
+				fieldKey,
+				mismatch: "fieldKind",
+				view: storedEmptyFieldSchema.kind,
+				stored: schema.kind,
+			} satisfies FieldKindDiscrepancy;
 		}
 	}
 }
+
 /**
- * @remarks
- *
  * This function uses incompatibilities to determine if changes to a document schema are backward-compatible, i.e., it determines
  * whether the `view` schema allows a superset of the documents that the `stored` schema allows.
+ *
+ * @remarks
  * According to the policy of schema evolution, `isRepoSuperset` supports three types of changes:
  * 1. Adding an optional field to an object node.
  * 2. Expanding the set of allowed types for a field.
  * 3. Relaxing a field kind to a more general field kind.
  *
  * Notes: We expect isRepoSuperset to return consistent results with allowsRepoSuperset. However, currently there are some scenarios
  * where the inconsistency will occur:
- *
- * - Different Node Kinds: If a and b have different node kinds (e.g., a is an objectNodeSchema and b is a mapNodeSchema),
- * `isRepoSuperset` will determine that a can never be the superset of b. In contrast, `allowsRepoSuperset` will continue
+ * - Different Node Kinds: If view and stored have different node kinds (e.g., view is an objectNodeSchema and stored is a mapNodeSchema),
+ * `isRepoSuperset` will determine that view can never be the superset of stored. In contrast, `allowsRepoSuperset` will continue
  * validating internal fields.
+ *
+ * TODO: Evaluate if this function is needed at all. It is only used in tests and could possibly be replaced with `allowsRepoSuperset`.
  */
-export function isRepoSuperset(view: TreeStoredSchema, stored: TreeStoredSchema): boolean {
+export function isViewSupersetOfStored(view: FieldSchema, stored: TreeStoredSchema): boolean {
```

Comment:
I don't actually understand the intended semantics of this function from its name.

Looking at the usage, I can't tell from that either as there is no production usage.

I think this function is supposed to return true if and only if you can upgrade stored to view?

If so, it seems really odd that the production code to compute canUpgrade doesn't use this: that fact seems to make this function entirely redundant / unnecessary. Its only use sees to be that its a second implementation of that logic we can compare to the production on in tests?